File: Substring with Concatenation of All Words.java
public class Solution {
    private boolean check(String S, String[] L) {
        boolean[] test = new boolean[L.length];
        int wlen = L[0].length();
        for (int i = 0; i < S.length(); i += wlen) {
            String sub = S.substring(i, i+wlen);
            boolean found = false;
            for (int j = 0; j < L.length; j++) {
                if (!test[j] && L[j].equals(sub)) {
                    found = true;
                    test[j] = true;
                    break;
                }
            }
            if (!found)
                return false;
        }
        return true;
    }
    public List<Integer> findSubstring(String S, String[] L) {
        List<Integer> ret = new ArrayList<Integer>();
        if (L.length == 0) {
            for (int i = 0; i < S.length(); i++)
                ret.add(i);
            return ret;
        }
        int wlen = L[0].length();
        int len = wlen * L.length;
        for (int i = 0; i < S.length(); i++) {
            if (i + len > S.length())
                break;
            if (check(S.substring(i, i + len), L))
                ret.add(i);
        }
        return ret;
    }
}

File: Substring with Concatenation of All Words.py
class Solution:
	def check(self, S, L):
		wordLen = len(L[0])
		strLen = wordLen * len(L)
		test = {}
		for i in xrange(len(L)):
			if (not L[i] in test):
				test[L[i]] = 1
			else:
				test[L[i]] = test[L[i]] + 1
		for i in range(0, strLen, wordLen):
			sub = S[i:(i+wordLen)]
			if (not sub in test):
				return False
			test[sub] = test[sub] - 1
			if (test[sub] < 0):
				return False
		return True
	# @param S, a string
	# @param L, a list of string
	# @return a list of integer
	def findSubstring(self, S, L):
		ret = []
		wordLen = len(L[0])
		strLen = wordLen * len(L)
		for i in xrange(len(S)):
			if (i + strLen > len(S)):
				break;
			if (self.check(S[i:(i+strLen)], L)):
				ret.append(i)
		return ret